urle381b6cba7_IkarosKappler_ikrs_http_tgz-pJ8-ikrs_json_rpc_JSONRPCHandlerJ8/src/ikrs/yuccasrv/socketmngr/ServerSocketThread.java

```diff
@@ -0,0 +1,468 @@
+				   char[] password,
+				   String keyStoreType ) 
+	throws KeyStoreException,
+	       FileNotFoundException,
+	       IOException,
+	       NoSuchAlgorithmException,
+	       CertificateException {
+
+	if( keyStoreType == null )
+	    keyStoreType = "JKS";
+
+	/*this.logger.log( Level.INFO,
+			   getClass().getName() + ".createSSLServerSocket(...)",
+			   "Initializing key store '" + fileName + "' ... " );
+	*/
+	KeyStore keyStore = KeyStore.getInstance( keyStoreType );
+	keyStore.load( new FileInputStream(fileName), 
+		       ( password == null ? new char[0] : password )
+		       ); 
+    
+	return keyStore;
+    }
+
+    public UUID getUUID() {
+	return this.uuid;
+    }
+
+    public InetAddress getBindAddress() {
+	return this.bindAddress;
+    }
+
+    public int getBindPort() {
+	return this.bindPort;
+    }
+
+    public Map<String,BasicType> getServerSettings() {
+	return this.bindSettings;
+    }
+    
+
+    public void run() {
+	
+	try {
+	    while( !this.isInterrupted() && !this.serverSocket.isClosed() ) {
+
+		Socket sock = this.serverSocket.accept();
+		    
+		// Increase connection counter
+		BasicType count = this.bindSettings.get(Constants.KEY_CONNECTION_COUNT);
+		if( count == null )
+		    count = new BasicNumberType( 1 );
+		else
+		    count = new BasicNumberType( count.getInt() + 1 );
+		this.bindSettings.put( Constants.KEY_CONNECTION_COUNT, count );
+		    
+		//System.out.println( getClass().getName() + ".run() Telling observer about incoming TCP connection ..." );
+		this.observer.incomingTCPConnection( this, sock );
+		//System.out.println( getClass().getName() + ".run() Done." );
+		
+		
+	    }
+	    // END while: server was closed or thread was interrupted
+
+	    if( !this.serverSocket.isClosed() ) 
+		closeServerSocket();
+	    else //if( !this.isInterrupted() )
+		this.observer.serverSocketClosed( this );
+
+	} catch( SocketTimeoutException e ) {
+
+	    if( !this.isInterrupted() )
+		this.observer.serverSocketException( this, e );
+
+	} catch( IOException e ) {
+	    
+	    if( !this.isInterrupted() )
+		this.observer.serverSocketException( this, e );
+
+	} catch( SecurityException e ) {
+	    
+	    if( !this.isInterrupted() )
+		this.observer.serverSocketException( this, e );
+
+	} catch( IllegalBlockingModeException e ) {
+
+	    if( !this.isInterrupted() )
+		this.observer.serverSocketException( this, e );
+	    
+	}
+	
+    }
+
+    public void interrupt() {
+
+	super.interrupt();
+
+	try {
+	    closeServerSocket();
+	} catch( IOException e ) {
+	    // Ignore
+	}
+
+    }
+
+    private void closeServerSocket() 
+	throws IOException {
+	if( this.serverSocket.isClosed() )
+	    return; // false;
+
+	//try {
+
+	    this.serverSocket.close();
+	    //this.observer.serverSocketClosed( this );
+	    //return true;
+	    
+	    //} catch( IOException e ) {
+	    
+	    //this.observer.serverSocketException( this, e );
+	    //return false;
+
+	    //}
+    }
+
+    public void finalize() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: This class extends the *Thread* superclass. If you do not invoke the finalizer on the superclass, its resources may not be disposed properly. Even when the superclass does not override the finalize() method, it is a best practice to invoke super.finalize() to prevent bugs in case the class hierarchy is changed in the future.

**Fix**: Invoke super.finalize() at the end of this method explicitly.